File: clients/typescript/src/stream.ts
// Streaming helpers: decode the response body's ReadableStream into lines, then
// surface SSE `data:` frames (chat streaming) or NDJSON objects (batch results).
//
// These are runtime-agnostic: they only use the WHATWG ReadableStream + TextDecoder
// that Node 20+, Deno, Bun and browsers all provide.

/** Yield decoded text lines from a byte stream, handling \n / \r\n splits across chunks. */
async function* iterLines(
  body: ReadableStream<Uint8Array>,
): AsyncGenerator<string, void, unknown> {
  const reader = body.getReader();
  const decoder = new TextDecoder("utf-8");
  let buffer = "";
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let nl: number;
      while ((nl = buffer.indexOf("\n")) !== -1) {
        let line = buffer.slice(0, nl);
        buffer = buffer.slice(nl + 1);
        if (line.endsWith("\r")) line = line.slice(0, -1);
        yield line;
      }
    }
    buffer += decoder.decode();
    if (buffer.endsWith("\r")) buffer = buffer.slice(0, -1);
    if (buffer.length > 0) yield buffer;
  } finally {
    reader.releaseLock();
  }
}

/**
 * Parse a `text/event-stream` body into the raw JSON payload of each `data:` frame.
 *
 * Stops (returns) on the sentinel line `data: [DONE]` WITHOUT yielding it — callers
 * must not JSON-parse the sentinel (SPEC.md). Blank lines (frame separators) and any
 * non-`data:` SSE fields (`event:`, `id:`, comments) are ignored. Multi-line `data:`
 * frames are concatenated with newlines per the SSE spec.
 */
export async function* parseSseData(
  body: ReadableStream<Uint8Array>,
): AsyncGenerator<string, void, unknown> {
  let dataLines: string[] = [];
  for await (const line of iterLines(body)) {
    if (line === "") {
      // Dispatch the accumulated event.
      if (dataLines.length > 0) {
        yield dataLines.join("\n");
        dataLines = [];
      }
      continue;
    }
    if (line.startsWith(":")) continue; // comment
    if (!line.startsWith("data:")) continue; // ignore event:/id:/retry:
    // Strip "data:" and a single optional leading space.
    let value = line.slice(5);
    if (value.startsWith(" ")) value = value.slice(1);
    if (value === "[DONE]") return;
    dataLines.push(value);
  }
  // Flush a trailing event that wasn't followed by a blank line.
  if (dataLines.length > 0) {
    const payload = dataLines.join("\n");
    if (payload !== "[DONE]") yield payload;
  }
}

/**
 * Parse an `application/x-ndjson` body into one parsed JSON value per non-empty line.
 */
export async function* parseNdjson(
  body: ReadableStream<Uint8Array>,
): AsyncGenerator<unknown, void, unknown> {
  for await (const line of iterLines(body)) {
    const trimmed = line.trim();
    if (trimmed === "") continue;
    yield JSON.parse(trimmed);
  }
}

File: clients/go/stream.go
package llmleaf

import (
	"bufio"
	"bytes"
	"encoding/json"
	"io"
	"net/http"
	"strings"

	pb "github.com/codefionn/llmleaf/clients/go/llmleafpb"
)

// ChatStream is an iterator over a streaming chat completion's
// ChatCompletionChunk frames. Call Recv until it returns io.EOF, then Close (or
// just defer Close). It is not safe for concurrent use.
type ChatStream struct {
	resp    *http.Response
	scanner *bufio.Scanner
	done    bool
	err     error
}

func newChatStream(resp *http.Response) *ChatStream {
	sc := bufio.NewScanner(resp.Body)
	// Allow large SSE frames (default 64 KiB max token; lift to 1 MiB).
	sc.Buffer(make([]byte, 0, 64*1024), 1<<20)
	return &ChatStream{resp: resp, scanner: sc}
}

// Recv returns the next chunk. It returns io.EOF once the stream is exhausted
// (after the `data: [DONE]` sentinel or EOF) and a non-EOF error on transport or
// decode failure. After any error (including io.EOF) the stream is finished.
func (s *ChatStream) Recv() (*pb.ChatCompletionChunk, error) {
	if s.err != nil {
		return nil, s.err
	}
	if s.done {
		return nil, io.EOF
	}

	var data bytes.Buffer
	haveData := false

	for s.scanner.Scan() {
		line := s.scanner.Text()

		// A blank line terminates the current SSE event.
		if line == "" {
			if !haveData {
				continue // stray blank line between events
			}
			payload := strings.TrimRight(data.String(), "\n")
			data.Reset()
			haveData = false

			if payload == "[DONE]" {
				s.finish(io.EOF)
				return nil, io.EOF
			}
			var wc wireChunk
			if err := json.Unmarshal([]byte(payload), &wc); err != nil {
				s.finish(err)
				return nil, err
			}
			return wc.toPB(), nil
		}

		// SSE comment lines start with ':'.
		if strings.HasPrefix(line, ":") {
			continue
		}

		field, value, found := strings.Cut(line, ":")
		if !found {
			field, value = line, ""
		} else if strings.HasPrefix(value, " ") {
			value = value[1:] // strip a single leading space per the SSE spec
		}
		if field == "data" {
			if value == "[DONE]" && !haveData {
				// Sentinel may arrive without a trailing blank line.
				s.finish(io.EOF)
				return nil, io.EOF
			}
			data.WriteString(value)
			data.WriteByte('\n')
			haveData = true
		}
		// Other SSE fields (event, id, retry) are ignored.
	}

	// Scanner exhausted. Flush a final event that lacked a trailing blank line.
	if err := s.scanner.Err(); err != nil {
		s.finish(err)
		return nil, err
	}
	if haveData {
		payload := strings.TrimRight(data.String(), "\n")
		if payload != "" && payload != "[DONE]" {
			var wc wireChunk
			if err := json.Unmarshal([]byte(payload), &wc); err != nil {
				s.finish(err)
				return nil, err
			}
			s.finish(io.EOF)
			return wc.toPB(), nil
		}
	}
	s.finish(io.EOF)
	return nil, io.EOF
}

func (s *ChatStream) finish(err error) {
	s.done = true
	if s.err == nil {
		s.err = err
	}
	s.resp.Body.Close()
}

// Close releases the underlying connection. Safe to call multiple times and
// after Recv has returned io.EOF.
func (s *ChatStream) Close() error {
	s.done = true
	return s.resp.Body.Close()
}
